Closes #7246  3.19 - Preconnect to external domains - Beacon insertion part

Khadreal opened the pull request:
# Description

Fixes #7246 

## Type of change

- [x] New feature (non-breaking change which adds functionality).

## Detailed scenario
N/A

## Technical description

### Documentation
Add factory class for Preload fonts to service provider

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

Khadreal committed:
:feat: Add factory class for service provider :closes: #7246

# Conflicts:
#	inc/Engine/Media/PreconnectExternalDomains/Database/Queries/PreconnectExternalDomains.php
#	inc/Engine/Media/PreconnectExternalDomains/ServiceProvider.php
#	inc/Plugin.php

Khadreal committed:
Add context for preconnect external domains

# Conflicts:
#	inc/Engine/Media/PreconnectExternalDomains/Context/Context.php
#	inc/Engine/Media/PreconnectExternalDomains/ServiceProvider.php

Khadreal committed:
Fixed php stan error -- #7246

# Conflicts:
#	inc/Engine/Media/PreconnectExternalDomains/ServiceProvider.php

Khadreal commented:
Test is failing which would be resolve when [this](https://github.com/wp-media/wp-rocket/issues/7248)  issue is completed.

Khadreal committed:
:feat: Add preconnect external domains db structure :closes: #7248

Khadreal committed:
:chore: Add ajax controller

Khadreal committed:
Add subscriber, table to uninstallation process

# Conflicts:
#	inc/Engine/Media/PreconnectExternalDomains/ServiceProvider.php

Miraeld commented:
Changes seems ok to me, however need to fix the tests if it's possible @Khadreal

Khadreal committed:
Add trait, update test

Khadreal committed:
Add unit test

Khadreal committed:
Update integration test setup

Khadreal committed:
Update feature name

# Conflicts:
#	inc/Engine/Common/PerformanceHints/Activation/ServiceProvider.php
#	inc/Engine/Common/PerformanceHints/ServiceProvider.php
#	inc/Engine/Media/PreconnectExternalDomains/Factory.php
#	inc/Engine/Media/PreconnectExternalDomains/ServiceProvider.php

MathieuLamiot commented:
@Khadreal @Miraeld Should this go back to In Progress then?
If this needs another PR to fix the test, please write it down here so we can know what needs to be done.

Khadreal commented:
> fix the test, please write it down

Left a comment here https://github.com/wp-media/wp-rocket/pull/7301#issuecomment-2656562555, the tests failing are dependent on the #7248 PR

Khadreal committed:
fixed php stan issue

Khadreal committed:
Add missing context class, subscriber class

Khadreal committed:
Add unit test for context

Khadreal committed:
Add more test to improve codacy score

Khadreal committed:
Add table, ajax controller to factory

Khadreal committed:
Add missing constructor argument

Khadreal committed:
Add traits for query filter :sparkles:

Khadreal committed:
Add test, fix php stan error

Khadreal committed:
uncomment test data, fix lint

Khadreal committed:
remove debugging code

Khadreal committed:
clean up, add row and schema table

Khadreal committed:
Merge branch 'feature/7248-data-insertion-part' into feature/7246-beacon-insertion-part

Khadreal committed:
fixed code conflict

hanna-meda reviewed (commented):
Thank you, @Khadreal, for this PR.
Tested on latest `feature/3.19` and everything is working as expected.
Related TP report below:
[testrail-report-714.pdf](https://github.com/user-attachments/files/20103355/testrail-report-714.pdf)
